src/util/lifecycle_utils.js

```diff
@@ -69,7 +69,154 @@ function file_matches_filter({obj_info, filter_func = undefined}) {
     return true;
 }
 
+/**
+ * get_lifecycle_rule_for_object determines the most specific matching lifecycle rule for the given object metadata
+ *
+ * priority is based on:
+ *  - longest matching prefix
+ *  - most matching tags
+ *  - narrowest object size range
+ *
+ * @param {Array<Object>} rules
+ * @param {Object} object_info
+ * @returns {Object|undefined}
+ */
+function get_lifecycle_rule_for_object(rules, object_info) {
+    let matched_rule;
+    let rule_priority = {
+        prefix_len: -1,
+        tag_count: -1,
+        size_span: Infinity,
+    };
+    if (object_info?.key && rules?.length > 0) {
+        for (const rule of rules) {
+            if (rule?.status !== 'Enabled') continue;
+
+            const filter = rule?.filter || {};
+
+            const filter_func = build_lifecycle_filter(filter);
+
+            if (!filter_func(object_info)) { continue; }
+
+            const priority = {
+                prefix_len: (filter?.prefix || '').length,
+                tag_count: Array.isArray(filter?.tags) ? filter?.tags.length : 0,
+                size_span: (filter?.object_size_less_than ?? Infinity) - (filter?.object_size_greater_than ?? 0)
+            };
+
+            // compare prefix length
+            const is_more_specific_prefix = priority.prefix_len > rule_priority.prefix_len;
+
+            // compare tag count (if prefixes are equal)
+            const is_more_specific_tags = priority.prefix_len === rule_priority.prefix_len &&
+                                        priority.tag_count > rule_priority.tag_count;
+
+            // compare size span (if prefixes and tags are equal)
+            const is_more_specific_size = priority.prefix_len === rule_priority.prefix_len &&
+                                        priority.tag_count === rule_priority.tag_count &&
+                                        priority.size_span < rule_priority.size_span;
+
+            if (is_more_specific_prefix || is_more_specific_tags || is_more_specific_size) {
```

Comment:
is there a priority between the conditions? what if one condition has more specific prefix, and the other more specific tags?

Reply:
Created a new function for priority. Thanks